mlir/lib/Transforms/Utils/LoopInvariantCodeMotionUtils.cpp

```diff
@@ -110,7 +110,7 @@ size_t mlir::moveLoopInvariantCode(LoopLikeOpInterface loopLike) {
         return loopLike.isDefinedOutsideOfLoop(value);
       },
       [&](Operation *op, Region *) {
-        return isMemoryEffectFree(op) && isSpeculatable(op);
+        return isMemoryEffectMovable(op) && isSpeculatable(op);
```

Comment:
This seems pretty expensive actually? Have you tried thinking about the complexity of the algorithm here?